packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts

```diff
@@ -500,84 +567,196 @@ export class CheckpointInfoIterationNode extends CheckpointInfoNode {
-    if (
-      bp.enabled &&
-      bp.condition &&
-      bp.condition!.toLowerCase().indexOf(CHECKPOINT) >= 0
-    ) {
-      if (!checkpointService.canAddCheckpointNote()) {
-        overTheLimit = true;
-        breakpointsToRemove.push(bp);
-        continue;
-      }
+  for (const bp of breakpointsChangedEvent.changed) {
+    const breakpointId = (bp as any)._id;
+    if (bp.condition && bp.condition!.toLowerCase().indexOf(CHECKPOINT) >= 0) {
       if (bp instanceof vscode.SourceBreakpoint) {
+        const checkpointOverlayAction = parseCheckpointInfoFromBreakpoint(bp);
         const uri = code2ProtocolConverter(bp.location.uri);
-        const lineNumber = bp.location.range.start.line + 1; // need to add 1 since the lines are 0 based
-        if (breakpointUtil.canSetLineBreakpoint(uri, lineNumber)) {
-          const typeRef = breakpointUtil.getTopLevelTyperefForUri(uri);
-          const filename = uri.substring(uri.lastIndexOf('/') + 1);
-          const theNode = await checkpointService.getOrCreateCheckpointNode(
+        const filename = uri.substring(uri.lastIndexOf('/') + 1);
+        const theNode = checkpointService.returnCheckpointNodeIfAlreadyExists(
+          breakpointId
+        );
+        // If the node exists then update it
+        if (theNode) {
+          theNode.updateCheckpoint(
+            bp.enabled,
             uri,
             filename,
-            typeRef,
-            lineNumber
+            checkpointOverlayAction
           );
-          const errorMessage = theNode.getActionCommandFailureMessage();
-          if (errorMessage) {
-            breakpointsToRemove.push(bp);
-            vscode.window.showErrorMessage(
-              nls.localize('unable_to_create_checkpoint') +
-                ' Error: ' +
-                errorMessage
-            );
-          }
         } else {
-          // The user tried to set a breakpoint on line that didn't contain any Apex
-          breakpointsToRemove.push(bp);
-          vscode.window.showErrorMessage(
-            nls.localize('checkpoints_can_only_be_on_valid_apex_source'));
+          // else if the node didn't exist then create it
+          checkpointService.createCheckpointNode(
+            breakpointId,
+            bp.enabled,
+            uri,
+            filename,
+            checkpointOverlayAction
+          );
         }
+      } else {
+        // The breakpoint is no longer a SourceBreakpoint which means if it ever was a checkpoint it has to be deleted
+        checkpointService.deleteCheckpointNodeIfExists(breakpointId);
+      }
+    } else {
+      // The breakpoint is not a checkpoint any longer, call to delete it from the checkpoint list if it exists
+      checkpointService.deleteCheckpointNodeIfExists(breakpointId);
+    }
+  }
+
+  for (const bp of breakpointsChangedEvent.added) {
+    if (bp.condition && bp.condition!.toLowerCase().indexOf(CHECKPOINT) >= 0) {
+      if (bp instanceof vscode.SourceBreakpoint) {
```

Comment:
You can include this in the first if statement

Reply:
Changed that and the a similar if statement for the changed breakpoints